### ISSUE TITLE
Support error based checks

### ISSUE THREAD

**Issue by bradleyayers** [jsverify/jsverify#140]
When using mocha it would be useful to be able to run checks by using error throwing assertions from [chai](http://chaijs.com/).

Here's a hacked up example of what I mean:

    function prop() {
        var propArgs = Array.prototype.slice.call(arguments);
        var propFunc = propArgs.pop();

        function debugFunc() {
            var funcContext = this;
            var funcArgs = arguments;

            var result = new Promise(function (resolve) {
                resolve(propFunc.apply(funcContext, funcArgs));
            });

            return result.then(
                function (x) {
                    // jsverify expects the thenable returned in a property to return "true" or "false". However often
                    // if we're just using chai, we don't both returning anything. In that case we should return "true"
                    // to indicate there was no problem (a problem would be indicated by an exception).
                    return typeof x === "undefined" ? true : x;
                }, function (err) {
                    throw err;
                });
        }

        jsc.property.apply(jsc, propArgs.concat(debugFunc));
    }

Excuse the names of the variables please, I'm still getting my head around the nomenclature. Basically the idea is to assume `return true` from a check function if a boolean wasn't returned. In addition catch any exception (I use the existing functionality of Promise to achieve this) and return that as an error.

**Comment by phadej** [comment]
You can throw exceptions (i.e. use chai) already now. As long as you `return true` in the happy case. Or am I missing something?

**Comment by bradleyayers** [comment]
Yeah I'm not sure why I ended up thinking that exceptions weren't handled properly, in my attempt to provide a concrete simple example, everything seems good. I suppose all I'd like to see here then is that returning a value be optional, and if no value is returned assume everything was fine.

This would mean that you could write a check that either raises an exception, or returns undefined, and that interface would be understood by jsverify.

**Comment by phadej** [comment]
I want support `undefined` as successful value. For example:

```js
jsc.property("obviously incorrect", "nat", function (n) {
  n === n + 1;
});
```

will be silently accepted.

**Comment by bradleyayers** [comment]
I understand where you're coming from, but I suspect that in practice people will want to use an assertion library like chai.js, because they capable of give more helpful error messages in some cases (e.g. giving the actual values that are unequal, or giving diffs of complex objects), make some types of assertions easier, or just often a more appealing API e.g. http://chaijs.com/api/bdd/

I agree with you that failing by default when _not_ using an assertion library is the sane behaviour, but I challenge you on what the normal user behaviour would be. I think most people would be coming from something like mocha or qunit where using an assertion library is the default.

Using an assertion library also makes it easier to do _multiple_ checks of a property, and get sensible error messages when one of those fails, e.g.

    jsc.property("obviously incorrect", jsc.json, jsc.jscon, function (o1, o2) {
      assert.deepProperty(o1, "foo.bar.baz");
      assert.deepEqual(o1, o2);
    });

Basically my thesis is that people will generally want to use jsverify with an assertion library, and as such requiring a return value will become boiler plate.

**Comment by phadej** [comment]
```js
jsc.property("obviously incorrect", jsc.json, jsc.json, function (o1, o2) {
  assert.deepProperty(o1, "foo.bar.baz");
  assert.deepEqual(o1, o2);
  return true;
});
```
isn't much more inconvenient.

We could add a friendly hint in the case property-function returns `undefined`, for example:

```
jsc.check(jsc.forall("nat", function () {}))
Failed after 1 tests and 1 shrinks. rngState: 0078f5cdb63a1e0249; Counterexample: 0;  [ 0 ]
Property function always returns 'undefined', have you forgot 'return true;' when using assertions?
```
